File: src/Arrays/ContainerWithMostWater.java
package Arrays;

public class ContainerWithMostWater {
    public int maxArea(int[] height) {
        if (height.length == 2) return Math.min(height[0], height[1]);
        int i = 0, j = height.length - 1, max = 0;
        while (i < j) {
            int minHeight = Math.min(height[i], height[j]);
            max = Math.max(max, minHeight * (j - i));
            while (i < j && minHeight > height[i]) i++;
            while (i < j && minHeight > height[j]) j--;
        }
        return max;
    }
    /* not effective
     public int maxArea(int[] height) {
        if (height.length == 2) return Math.min(height[0], height[1]);
        int i = 0, j = height.length - 1, max = 0;
        while (i < j) {
            int subMax = Math.min(height[i], height[j]) * (j - i);
            if (subMax > max) {
                max = subMax;
            }
            if (height[i] > height[j])
                j--;
            else
                i++;
        }
        return max;
    }
     */
}

File: src/DailyLeetCodingChallenge/ContainerWithMostWater.kt
package DailyLeetCodingChallenge

import kotlin.math.max
import kotlin.math.min

fun maxArea(height: IntArray): Int {
    var maxSum = 0
    var l = 0
    var r = height.size - 1
    while (r > l) {
        maxSum = max(maxSum, min(height[r], height[l]) *  (r - l))
        if (height[r] > height[l]) l++
        else r--
    }
    return maxSum
}